### ISSUE TITLE
How pysolr import UTF-8 data to Solr server?

### ISSUE THREAD

**Issue by dungdb1** [django-haystack/pysolr#177]
I have a request look like this:
```
    solr.add(    [
        {
            "id": "1",
            "title": "đinh bộ linh",
            "content": ["ông bà "]
        }]
)
```

When i call this request. Having a error like this:
```
Traceback (most recent call last):
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\threading.py", line 914, in _
bootstrap_inner
    self.run()
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\threading.py", line 1180, in
run
    self.function(*self.args, **self.kwargs)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\mysql_replicati
on-0.7-py3.5.egg\pymysqlreplication\row_event.py", line 739, in AddThread
    solr.add(value["value"])
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pysolr-3.3.3-py
3.5.egg\pysolr.py", line 822, in add
    return self._update(m, commit=commit, softCommit=softCommit, waitFlush=waitFlush, waitSearcher=w
aitSearcher)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pysolr-3.3.3-py
3.5.egg\pysolr.py", line 400, in _update
    return self._send_request('post', path, message, {'Content-type': 'text/xml; charset=utf-8'})
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pysolr-3.3.3-py
3.5.egg\pysolr.py", line 309, in _send_request
    timeout=self.timeout)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\sessions.py", line 511, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\sessions.py", line 454, in request
    prep = self.prepare_request(req)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\sessions.py", line 388, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\models.py", line 308, in prepare
    self.prepare_body(data, files, json)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\models.py", line 459, in prepare_body
    body = self._encode_params(data)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\models.py", line 85, in _encode_params
    return to_native_string(data)
  File "C:\Users\dungdb1\AppData\Local\Programs\Python\Python35-32\lib\site-packages\requests-2.9.0-
py3.5.egg\requests\utils.py", line 702, in to_native_string
    out = string.decode(encoding)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 91: ordinal not in range(128)
```

how can you send that to solr using pysolr? Pls help me! Thanks

**Comment by joshuajonah** [comment]
+1

**Comment by acdha** [comment]
Do you have a failing test case which demonstrates your problem? Pysolr should handle provided unicode objects without issue – e.g. most of my content is non-ASCII – so it's not immediately obvious what is causing this.

**Comment by joshuajonah** [comment]
Traceback (most recent call last):
	  File "./manage.py", line 8, in <module>
	    execute_from_command_line(sys.argv)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
	    utility.execute()
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/__init__.py", line 345, in execute
	    self.fetch_command(subcommand).run_from_argv(self.argv)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/base.py", line 348, in run_from_argv
	    self.execute(*args, **cmd_options)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/base.py", line 399, in execute
	    output = self.handle(*args, **options)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/haystack/management/commands/rebuild_index.py", line 37, in handle
	    call_command('update_index', **options)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/__init__.py", line 119, in call_command
	    return command.execute(*args, **defaults)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/django/core/management/base.py", line 399, in execute
	    output = self.handle(*args, **options)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/haystack/management/commands/update_index.py", line 214, in handle
	    self.update_backend(label, using)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/haystack/management/commands/update_index.py", line 257, in update_backend
	    commit=self.commit, max_retries=self.max_retries)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/haystack/management/commands/update_index.py", line 84, in do_update
	    backend.update(index, current_qs, commit=commit)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/haystack/backends/solr_backend.py", line 72, in update
	    self.conn.add(docs, commit=commit, boost=index.get_field_weights())
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/pysolr.py", line 875, in add
	    overwrite=overwrite, handler=handler)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/pysolr.py", line 471, in _update
	    return self._send_request('post', path, message, {'Content-type': 'text/xml; charset=utf-8'})
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/pysolr.py", line 361, in _send_request
	    timeout=self.timeout)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/sessions.py", line 511, in post
	    return self.request('POST', url, data=data, json=json, **kwargs)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/sessions.py", line 454, in request
	    prep = self.prepare_request(req)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/sessions.py", line 388, in prepare_request
	    hooks=merge_hooks(request.hooks, self.hooks),
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/models.py", line 297, in prepare
	    self.prepare_body(data, files, json)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/models.py", line 448, in prepare_body
	    body = self._encode_params(data)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/models.py", line 85, in _encode_params
	    return to_native_string(data)
	  File "/home/joshua/Development/ncweb/lib/python3.4/site-packages/requests/utils.py", line 700, in to_native_string
	    out = string.decode(encoding)
	UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 336: ordinal not in range(128)

Well, it's trying to `.decode('ascii')` by default in requests. How do we get around this. A simple &reg; in the text is doing this.

**Comment by joshuajonah** [comment]
I've not done anything special, it's just a simple `models.TextField` using a simple `indexes.CharField`

**Comment by acdha** [comment]
Can you distill that into a program which reliably crashes? Without a way to reproduce what you're seeing it's much harder to guess about what might be going wrong.

**Comment by acdha** [comment]
That code is in requests – https://github.com/kennethreitz/requests/blob/master/requests/utils.py#L773-L786 – and it should only being called if you're passing in byte-strings rather than native unicode strings. That would depend on how your data is being prepared but it's where I would start looking.

**Comment by joshuajonah** [comment]
Ok, found where it is being forced to bytes. It appears to be by default in Py3: https://github.com/django-haystack/pysolr/blob/master/pysolr.py#L357

How are you NOT having this issue?

**Comment by acdha** [comment]
Please remember that I am volunteering my time to try to help you. I don't get paid to support pysolr or haystack and that means that anything you can do to make it easier for me to see what you're seeing dramatically increases the likelihood of us being able to fix it.

Both pysolr and haystack have test suites which pass on Python 3, even with Unicode content, and my own apps work on both, again with lots of non-ASCII text. That likely means that you're using a combination which isn't tested and should be. If you could contribute a test which fails or at least a full program which anyone else could run, that would make the process much easier. Even just having the `repr()` output for the data being passed in to e.g. `self.conn.add` might shed some light on what's going on. In my testing, requests and pysolr handle both unicode strings and encoded byte strings correctly so there's obviously some difference which is not immediately obvious.

**Comment by joshuajonah** [comment]
I did some research into fixing the issue with sending a UTF-8 body in requests only to find that my version was behind, it was fixed here: https://github.com/kennethreitz/requests/commit/b32c3bc10f0fb4075b6a282b283418f5290e812a

Feeling pretty silly here. If you have this issue anybody, update your outdated Requests.

**Comment by acdha** [comment]
Thanks! What version was that fix in? I'll update pysolr to depend on the newer release.

**Comment by joshuajonah** [comment]
That was fixed in 2.9.1
